### DISCUSSION TITLE
How to define multiple entry points for RsLib?

### DISCUSSION THREAD

**Discussion by 0x80** [web-infra-dev/rsbuild#3702] [Q&A] [answered]
I am attempting to migrate a project from tsup to rslib, but I don't know how to translate this:

```ts
import { defineConfig } from "tsup";

export default defineConfig({
  entry: {
    "utils/index": "src/utils/index.ts",
    firebase: "src/firebase.ts",
    firestore: "src/firestore.ts",
    hash: "src/hash.ts",
    redis: "src/redis.ts",
    tasks: "src/tasks.ts",
    timer: "src/timer.ts",
  },
  format: ["esm"],
  target: "es2022",
  sourcemap: true,
});
```

I can't seem to find anything about defining multiple entries in the docs or source code...

**Comment by chenjiahan** [comment, marked as answer]
Rslib can use the [source.entry](https://rsbuild.dev/config/source/entry) config of Rsbuild, here is an example:

```ts
export default defineConfig({
  lib: [
    {
      format: 'esm',
      source: {
        entry: {
          foo: './src/foo.ts',
        },
      },
    },
  ],
});
```

**Comment by 0x80** [comment]
Thanks that kind of worked. 

But it seems my exports are somehow not getting exported from each of the entries.

I was trying to replace tsup with rslib. Maybe I misunderstood the use-case. It doesn't appear to generate the clean ESM output I was hoping for.

Is it possible for rslib to output code without all the webpack indirections or is that just inherit to the goal of being a webpack alternative?

**Reply by chenjiahan** [reply]
Rslib aims to output clean ESM code without webpack runtimes, and this is still work in progress. If you find any bad cases, feel free to open an issue in the Rslib repo.
